### PR TITLE
NODE-1525 refactor, clean entire codebase

### PR DESCRIPTION
**Pull request by mbroadst**
This is the initial work of or kerberos refactor for Q2, and comprises a number of massive breaking changes for the sake of maintainability and stability of the module:

- The entire public API is changed, and is primarily comprised of three methods `initializeClient`, `initializeServer` and `checkPassword`
- The addon code is completely rewritten, with an eye towards providing the same API for both windows and unix builds of the module. All code is modernized here to use c++11.
- Windows support has been removed for this first step, and will be reintroduced in the next pull request
- The code is now linted both in javascript (eslint), and cxx (clang-format)

### CODE REVIEW

File: lib/auth_processes/mongodb.js [hunk 1]
```diff
@@ -140,208 +150,236 @@ Win32MongoProcessor.second_transition = function(self) {
...
     });
-  }
-}
+  };
+};
 
 Win32MongoProcessor.third_transition = function(self) {
   return function(payload, callback) {
-    var messageLength = 0;
+    const messageLength = 0;
```
Comment (daprahamian) [inline lib/auth_processes/mongodb.js:158]: Should this be `let`?

File: src/kerberos.cc [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+#define GSS_MECH_OID_KRB5 9
+#define GSS_MECH_OID_SPNEGO 6
+
+static char krb5_mech_oid_bytes[] = "\x2a\x86\x48\x86\xf7\x12\x01\x02\x02";
+gss_OID_desc krb5_mech_oid = {9, &krb5_mech_oid_bytes};
+
+static char spnego_mech_oid_bytes[] = "\x2b\x06\x01\x05\x05\x02";
+gss_OID_desc spnego_mech_oid = {6, &spnego_mech_oid_bytes};
+
```
Comment (daprahamian) [inline src/kerberos.cc:18]: Is it worth anonymous namespacing all the classes? Or is no one gonna pull this in like that?

**Review by daprahamian — APPROVED** [review]
👍 with a few questions

File: lib/auth_processes/mongodb.js (hunk 1 above)
Comment (mbroadst) [reply lib/auth_processes/mongodb.js:158]: this code is all going away, but yeah it should be

**Pull request merged**